File: lib_base/src/main/java/com/android/base/app/CrashHandler.java
package com.android.base.app;

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.Process;

import com.blankj.utilcode.util.AppUtils;

import java.io.File;
import java.io.PrintStream;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

import timber.log.Timber;

/**
 * 全局异常处理
 */
final class CrashHandler implements Thread.UncaughtExceptionHandler {

    private Context mContext;
    private Sword.CrashProcessor mCrashProcessor;

    public static CrashHandler register(Application application) {
        CrashHandler crashHandler = new CrashHandler(application);
        Thread.setDefaultUncaughtExceptionHandler(crashHandler);
        return crashHandler;
    }

    void setCrashProcessor(Sword.CrashProcessor crashProcessor) {
        mCrashProcessor = crashProcessor;
    }

    private CrashHandler(Context context) {
        this.mContext = context;
    }

    @Override
    public void uncaughtException(Thread thread, Throwable ex) {

        if (mCrashProcessor != null) {
            mCrashProcessor.uncaughtException(thread, ex);
        } else {
            // 收集异常信息，写入到sd卡
            restoreCrash(thread, ex);
            //退出
            killProcess();
        }
    }

    private void restoreCrash(@SuppressWarnings("unused") Thread thread, Throwable ex) {
        ex.printStackTrace(System.err);
        // 收集异常信息，写入到sd卡
        File dir = new File(mContext.getExternalFilesDir(null) + File.separator + "crash");

        if (!dir.exists()) {
            boolean mkdirs = dir.mkdirs();
            if (!mkdirs) {
                Timber.e("CrashHandler create dir fail");
                return;
            }
        }

        try {
            @SuppressLint("SimpleDateFormat")
            DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            String name = dateFormat.format(new Date(System.currentTimeMillis())) + ".log";
            File fileName = new File(dir, name);
            if (!fileName.exists()) {
                @SuppressWarnings("unused")
                boolean newFile = fileName.createNewFile();
            }

            PrintStream err = new PrintStream(fileName);

            err.println("--------------------------------AppInfo--------------------------------");
            err.println("AndroidVersion: " + AppUtils.getAppVersionName());
            err.println();
            err.println();
            err.println("--------------------------------SystemInfo:--------------------------------");
            err.println("Product: " + android.os.Build.PRODUCT);
            err.println("CPU_ABI: " + android.os.Build.CPU_ABI);
            err.println("TAGS: " + android.os.Build.TAGS);
            err.println("VERSION_CODES.BASE:" + android.os.Build.VERSION_CODES.BASE);
            err.println("MODEL: " + android.os.Build.MODEL);
            err.println("SDK: " + Build.VERSION.SDK_INT);
            err.println("VERSION.RELEASE: " + android.os.Build.VERSION.RELEASE);
            err.println("DEVICE: " + android.os.Build.DEVICE);
            err.println("DISPLAY: " + android.os.Build.DISPLAY);
            err.println("BRAND: " + android.os.Build.BRAND);
            err.println("BOARD: " + android.os.Build.BOARD);
            err.println("FINGERPRINT: " + android.os.Build.FINGERPRINT);
            err.println("ID: " + android.os.Build.ID);
            err.println("MANUFACTURER: " + android.os.Build.MANUFACTURER);
            err.println("USER: " + android.os.Build.USER);
            err.println();
            err.println();
            err.println("--------------------------------CrashContent--------------------------------");
            ex.printStackTrace(err);
            err.println();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void killProcess() {
        Process.killProcess(Process.myPid());
    }

}

File: lib_base/src/main/java/com/android/base/architecture/app/CrashHandler.kt
package com.android.base.architecture.app

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.os.Build
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Process
import com.android.base.CrashProcessor
import com.android.base.utils.android.AppUtils
import timber.log.Timber
import java.io.File
import java.io.PrintStream
import java.lang.Thread.UncaughtExceptionHandler
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

/**
 * 全局异常处理
 */
internal class CrashHandler private constructor(
        private val context: Context
) : UncaughtExceptionHandler {

    private var _crashProcessor: CrashProcessor? = null

    fun setCrashProcessor(crashProcessor: CrashProcessor) {
        _crashProcessor = crashProcessor
    }

    override fun uncaughtException(thread: Thread, ex: Throwable) {
        val crashProcessor = _crashProcessor
        if (crashProcessor != null) {
            crashProcessor.uncaughtException(thread, ex)
        } else {
            // 收集异常信息，写入到sd卡
            restoreCrash(thread, ex)
            //退出
            killProcess()
        }
    }

    private fun restoreCrash(thread: Thread, ex: Throwable) {
        ex.printStackTrace(System.err)
        // 收集异常信息，写入到sd卡
        val externalFilesDir = context.getExternalFilesDir(null) ?: return
        val dir = File(externalFilesDir.toString() + File.separator + "crash")

        if (!dir.exists()) {
            val mkdirs = dir.mkdirs()
            if (!mkdirs) {
                Timber.e("CrashHandler create dir fail")
                return
            }
        }

        try {
            @SuppressLint("SimpleDateFormat") val dateFormat: DateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
            val name = dateFormat.format(Date(System.currentTimeMillis())) + ".log"
            val targetFile = File(dir, name)
            if (!targetFile.exists()) {
                targetFile.createNewFile()
            }
            val err = PrintStream(targetFile)
            err.println("--------------------------------AppInfo--------------------------------")
            err.println("AndroidVersion: " + AppUtils.getAppVersionName())
            err.println()
            err.println()
            err.println("--------------------------------SystemInfo:--------------------------------")
            err.println("Product: " + Build.PRODUCT)
            err.println("CPU_ABI: " + Build.CPU_ABI)
            err.println("TAGS: " + Build.TAGS)
            err.println("VERSION_CODES.BASE:" + VERSION_CODES.BASE)
            err.println("MODEL: " + Build.MODEL)
            err.println("SDK: " + VERSION.SDK_INT)
            err.println("VERSION.RELEASE: " + VERSION.RELEASE)
            err.println("DEVICE: " + Build.DEVICE)
            err.println("DISPLAY: " + Build.DISPLAY)
            err.println("BRAND: " + Build.BRAND)
            err.println("BOARD: " + Build.BOARD)
            err.println("FINGERPRINT: " + Build.FINGERPRINT)
            err.println("ID: " + Build.ID)
            err.println("MANUFACTURER: " + Build.MANUFACTURER)
            err.println("USER: " + Build.USER)
            err.println()
            err.println()
            err.println("--------------------------------CrashContent--------------------------------")
            ex.printStackTrace(err)
            err.println()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun killProcess() {
        Process.killProcess(Process.myPid())
    }

    companion object {
        fun register(application: Application): CrashHandler {
            val crashHandler = CrashHandler(application)
            Thread.setDefaultUncaughtExceptionHandler(crashHandler)
            return crashHandler
        }
    }

}